drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1505 @@
+		LOG_INST_ERR(cfg->log, "%u, fail to clear interrupt, code:%d.", __LINE__, ret);
+		return -EHOSTDOWN;
+	}
+
+	ret = am_hal_mspi_interrupt_enable(data->mspiHandle,
+					   AM_HAL_MSPI_INT_CQUPD | AM_HAL_MSPI_INT_ERR);
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u, fail to turn on interrupt, code:%d.", __LINE__, ret);
+		return -EHOSTDOWN;
+	}
+
+	cfg->irq_cfg_func();
+
+	mspi_context_unlock_unconditionally(&data->ctx);
+
+	if (config->re_init) {
+		k_mutex_unlock(&data->lock);
+	}
+
+	return ret;
+}
+
+static int mspi_ambiq_dev_config(const struct device *controller, const struct mspi_dev_id *dev_id,
+				 const enum mspi_dev_cfg_mask param_mask,
+				 const struct mspi_dev_cfg *dev_cfg)
+{
+	const struct mspi_ambiq_config *cfg = controller->config;
+	struct mspi_ambiq_data *data = controller->data;
+	am_hal_mspi_dev_config_t hal_dev_cfg = data->hal_dev_cfg;
+	int ret = 0;
+
+	if (data->dev_id != dev_id) {
+		if (k_mutex_lock(&data->lock, K_MSEC(CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE))) {
+			LOG_INST_ERR(cfg->log, "%u, fail to gain controller access.", __LINE__);
+			return -EBUSY;
+		}
+
+		ret = mspi_verify_device(controller, dev_id);
+		if (ret) {
+			goto e_return;
+		}
+	}
+
+	if (mspi_is_inp(controller)) {
+		ret = -EBUSY;
+		goto e_return;
+	}
+
+	if (param_mask == MSPI_DEVICE_CONFIG_NONE && !cfg->mspicfg.sw_multi_periph) {
+		/* Do nothing except obtaining the controller lock */
+		data->dev_id = (struct mspi_dev_id *)dev_id;
+		return ret;
+
+	} else if (param_mask != MSPI_DEVICE_CONFIG_ALL) {
+		if (data->dev_id != dev_id) {
+			LOG_INST_ERR(cfg->log, "%u, config failed, must be the same device.",
+				     __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if ((param_mask & (~(MSPI_DEVICE_CONFIG_FREQUENCY | MSPI_DEVICE_CONFIG_IO_MODE |
+				     MSPI_DEVICE_CONFIG_CE_NUM | MSPI_DEVICE_CONFIG_DATA_RATE |
+				     MSPI_DEVICE_CONFIG_CMD_LEN | MSPI_DEVICE_CONFIG_ADDR_LEN)))) {
+			LOG_INST_ERR(cfg->log, "%u, config type not supported.", __LINE__);
+			ret = -ENOTSUP;
+			goto e_return;
+		}
+
+		if (param_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+			hal_dev_cfg.eClockFreq = mspi_set_freq(cfg, dev_cfg->freq);
+			if (hal_dev_cfg.eClockFreq == 0) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+			ret = am_hal_mspi_control(data->mspiHandle, AM_HAL_MSPI_REQ_CLOCK_CONFIG,
+						  &hal_dev_cfg.eClockFreq);
+			if (ret) {
+				LOG_INST_ERR(cfg->log, "%u, failed to configure eClockFreq.",
+					     __LINE__);
+				ret = -EHOSTDOWN;
+				goto e_return;
+			}
+			data->dev_cfg.freq = dev_cfg->freq;
+		}
+
+		if ((param_mask & MSPI_DEVICE_CONFIG_IO_MODE) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_CE_NUM) ||
+		    (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE)) {
+			hal_dev_cfg.eDeviceConfig = mspi_set_line(
+				cfg, dev_cfg->io_mode, dev_cfg->data_rate, dev_cfg->ce_num);
+			if (hal_dev_cfg.eDeviceConfig == AM_HAL_MSPI_FLASH_MAX) {
+				ret = -ENOTSUP;
+				goto e_return;
+			}
+
+			ret = am_hal_mspi_disable(data->mspiHandle);
```

Comment:
should be removed.
The reason is that disable/enable sequence is going to reset the command queue. Unlike MSPI_DEVICE_CONFIG_ALL, all other supported type meant to be a quick in and out configuration and thus should not tamper the hal states.